src/vs/platform/extensionManagement/common/extensionsScannerService.ts

```diff
@@ -0,0 +1,658 @@
+		let extensions: IRelaxedScannedExtension[];
+		if (scanOptions.includeUninstalled) {
+			extensions = await this.scanFromUserExtensionsLocation();
+		} else {
+			let [uninstalled, scannedExtensions] = await Promise.all([this.getUninstalledExtensions(), this.scanFromUserExtensionsLocation()]);
+			extensions = scannedExtensions.filter(e => !uninstalled[ExtensionKey.create(e).toString()]);
+		}
+		extensions = await this.applyScanOptions(extensions, scanOptions, true);
+		this.logService.trace('Scanned user extensions:', extensions.length);
+		return extensions;
+	}
+
+	async scanExtensionsUnderDevelopment(scanOptions: ScanOptions): Promise<IScannedExtension[]> {
+		if (this.environmentService.isExtensionDevelopment && this.environmentService.extensionDevelopmentLocationURI) {
+			const extensions = (await Promise.all(this.environmentService.extensionDevelopmentLocationURI.filter(extLoc => extLoc.scheme === Schemas.file)
+				.map(async extensionDevelopmentLocationURI => await this.doScanOneOrMultipleExtensions(extensionDevelopmentLocationURI, ExtensionType.User))))
+				.flat();
+			return this.applyScanOptions(extensions, scanOptions, true);
+		}
+		return [];
+	}
+
+	async scanExistingExtension(extensionLocation: URI, extensionType: ExtensionType, scanOptions: ScanOptions): Promise<IScannedExtension | null> {
+		const extension = await this.scanExtension(extensionLocation, extensionType);
+		if (!extension) {
+			return null;
+		}
+		if (!scanOptions.includeInvalid && !extension.isValid) {
+			return null;
+		}
+		extension.manifest = await this.localizeManifest(extension.location, extension.manifest, scanOptions.nlsConfiguration);
+		return extension;
+	}
+
+	async scanOneOrMultipleExtensions(extensionLocation: URI, extensionType: ExtensionType, scanOptions: ScanOptions): Promise<IScannedExtension[]> {
+		const extensions = await this.doScanOneOrMultipleExtensions(extensionLocation, extensionType);
+		return this.applyScanOptions(extensions, scanOptions, true);
+	}
+
+	async updateMetadata(extensionLocation: URI, metaData: Partial<Metadata>): Promise<void> {
+		const manifest = await this.scanExtensionManifest(extensionLocation);
+		if (!manifest) {
+			throw new Error(localize('cannot read', "Cannot read the extension manifest from {0}", extensionLocation.path));
+		}
+		// unset if false
+		metaData.isMachineScoped = metaData.isMachineScoped || undefined;
+		metaData.isBuiltin = metaData.isBuiltin || undefined;
+		metaData.installedTimestamp = metaData.installedTimestamp || undefined;
+		manifest.__metadata = { ...manifest.__metadata, ...metaData };
+		await this.fileService.writeFile(joinPath(extensionLocation, 'package.json'), VSBuffer.fromString(JSON.stringify(manifest, null, '\t')));
+	}
+
+	private async applyScanOptions(extensions: IRelaxedScannedExtension[], scanOptions: ScanOptions, pickLatest: boolean): Promise<IRelaxedScannedExtension[]> {
+		if (!scanOptions.includeAllVersions) {
```

Comment:
I think this was one of the edge cases we discussed about. If there are two versions of the same extension on disk, `ext-1.0.0` which wants vscode 1.66 and `ext-2.0.0` which wants vscode 1.67, when scanning with vscode 1.66, it would be preferable to return `ext-1.0.0` and when scanning with vscode 1.67, it would be preferable to return `ext-2.0.0`. I believe the current code would not load the extension when using 1.66.

Reply:
This case is considered here:

https://github.com/microsoft/vscode/blob/9ecfc9365fcbafe3538b2c978f2e18be951c4c79/src/vs/platform/extensionManagement/common/extensionsScannerService.ts#L328-L340

I also have a unit test for this case:

https://github.com/microsoft/vscode/blob/9fe18d2e6b39e321fa9c0d780a30734b1366546f/src/vs/platform/extensionManagement/test/node/extensionsScannerService.test.ts#L161-L171

Reply:
👏 You're right! Now I saw that `isValid` is explicitly used in `dedupExtensions`!